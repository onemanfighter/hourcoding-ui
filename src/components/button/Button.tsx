import { ButtonProps, ButtonVariant } from './types';
import { motion } from 'framer-motion';
import { useHcColor, useHcComponentTheme, useHcTextVariants } from '../../theme';
import { BUTTON_BORDER_RADIUS } from './constants';

const Button = ({
  size = 'large',
  label,
  isDisabled = false,
  variant = ButtonVariant.CONTAINED,
  clickHandler,
}: ButtonProps) => {
  const buttonTheme = useHcComponentTheme().Button[variant];
  const getTextVariant = useHcTextVariants();
  const getColor = useHcColor();

  const textVariant = () => {
    if (size === 'small') {
      return getTextVariant('buttonSmall');
    } else if (size === 'default') {
      return getTextVariant('buttonDefault');
    } else if (size === 'large') {
      return getTextVariant('buttonLarge');
    } else {
      return getTextVariant('buttonDefault');
    }
  };

  const onClickHandler = () => {
    if (isDisabled) {
      return;
    }

    if (clickHandler) clickHandler();
  };

  return (
    <motion.button
      onClick={onClickHandler}
      disabled={isDisabled}
      animate={{
        scale: isDisabled ? 1 : 1.1,
        transition: {
          duration: 0.3,
        },
      }}
      whileHover={
        isDisabled
          ? {}
          : {
              scale: 1.2,
              backgroundColor: getColor('neutral.600'),
              transition: {
                duration: 0.3,
              },
            }
      }
      whileTap={
        isDisabled
          ? {}
          : {
              scale: 1,
              transition: {
                duration: 0.1,
              },
            }
      }
      style={{
        padding: size === 'small' ? '4px 8px' : size === 'default' ? '6px 12px' : '8px 16px',
        border: variant === ButtonVariant.OUTLINED ? '1px solid' : 'none',
        ...textVariant(),
        borderRadius: BUTTON_BORDER_RADIUS,
        borderColor: buttonTheme.borderColor ? getColor(buttonTheme.borderColor) : 'transparent',
        color: getColor(buttonTheme.color),
        background: buttonTheme.backgroundColor
          ? getColor(buttonTheme.backgroundColor)
          : 'transparent',
      }}
    >
      {label}
    </motion.button>
  );
};

export default Button;
