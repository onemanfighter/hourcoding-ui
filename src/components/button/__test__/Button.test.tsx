import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Button from '../Button';

describe('Login component tests', () => {
  beforeEach(() => {
    // write someting before each test
  });

  afterEach(() => {
    // write someting after each test
  });

  it('Renders correctly initial document', async () => {
    /* first we visit /login and test if the string in the element with class "login-label"  has"Please Log In" is there */
    render(<Button variant="contained" label="Click me" size="small" />);
    const loginLabel = screen.getByText('Click me');

    expect(loginLabel).toBeDefined();
  });

  it('Snapshot test', () => {
    const tree = renderer
      .create(<Button variant="contained" label="Click me" size="small" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
