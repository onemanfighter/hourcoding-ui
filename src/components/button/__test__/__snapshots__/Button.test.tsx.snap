// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button contained snapshot test 1`] = `
<button
  disabled={false}
  onClick={[Function]}
  style={
    {
      "background": "#86efac",
      "border": "none",
      "borderColor": "#a4cafe",
      "borderRadius": 4,
      "color": undefined,
      "fontSize": 14,
      "fontWeight": "400",
      "lineHeight": 1.5,
      "padding": "4px 8px",
    }
  }
  tabIndex={0}
>
  Click me
</button>
`;
